File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/search/ResultAdapter.java
package com.yichiuan.moedict.ui.search;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.yichiuan.moedict.R;
import com.yichiuan.moedict.ui.main.MainActivity;

import java.util.ArrayList;

import moe.Index;

public class ResultAdapter extends RecyclerView.Adapter<ResultAdapter.ResultViewHolder>
        implements View.OnClickListener {

    private LayoutInflater inflater;

    private Index index;

    private ArrayList<Integer> results;

    ResultAdapter(Context context, Index index) {
        inflater = LayoutInflater.from(context);
        this.index = index;
    }

    void setResults(ArrayList<Integer> results) {
        this.results = results;
        notifyDataSetChanged();
    }

    ArrayList<Integer> getResults() {
        return results;
    }

    @Override
    public ResultViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ResultViewHolder holder = new ResultViewHolder(
                inflater.inflate(R.layout.item_search_result, parent, false));
        holder.itemView.setOnClickListener(this);
        return holder;
    }

    @Override
    public void onBindViewHolder(ResultViewHolder holder, int position) {
        String word = index.words(results.get(position));
        holder.item.setText(word);
        holder.itemView.setTag(word);
    }

    @Override
    public int getItemCount() {
        return results.size();
    }

    @Override
    public void onClick(View view) {
        Context context = inflater.getContext();

        Intent show = new Intent(context, MainActivity.class);
        show.putExtra(SearchManager.QUERY, (String)view.getTag());
        context.startActivity(show);
    }

    static class ResultViewHolder extends RecyclerView.ViewHolder {

        TextView item;

        public ResultViewHolder(View itemView) {
            super(itemView);
            item = itemView.findViewById(R.id.textview_item_result);
        }
    }
}

File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/search/ResultAdapter.kt
package com.yichiuan.moedict.ui.search

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.yichiuan.moedict.R
import com.yichiuan.moedict.ui.main.MainActivity
import java.util.*

class ResultAdapter internal constructor(context: Context, private val model: SearchViewModel
) : RecyclerView.Adapter<ResultAdapter.ResultViewHolder>(), View.OnClickListener {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    internal var results: ArrayList<Int>? = null
        set(results) {
            field = results
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ResultViewHolder {
        val holder = ResultViewHolder(
                inflater.inflate(R.layout.item_search_result, parent, false))
        holder.itemView.setOnClickListener(this)
        return holder
    }

    override fun onBindViewHolder(holder: ResultViewHolder, position: Int) {
        val word = model.getWord(this.results!![position])
        holder.item.text = word
        holder.itemView.tag = word
    }

    override fun getItemCount(): Int {
        return this.results!!.size
    }

    override fun onClick(view: View) {
        val context = inflater.context

        val show = Intent(context, MainActivity::class.java)
        show.putExtra(SearchManager.QUERY, view.tag as String)
        context.startActivity(show)
    }

    class ResultViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var item: TextView = itemView.findViewById(R.id.textview_item_result)
    }
}
